File: 86. Partition List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def partition(self, head: ListNode, x: int) -> ListNode:
        if not head or not head.next:
            return head
        lh = ListNode(0)
        gh = ListNode(0)
        lc = lh
        gc = gh
        while head:
            if head.val < x:
                lc.next = head
                lc = lc.next
            else:
                gc.next = head
                gc = gc.next
            head = head.next
        gc.next = None
        lc.next = gh.next
        return lh.next

File: 86. Partition List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode dummy1 = new ListNode(0);
        ListNode dummy2 = new ListNode(0);
        ListNode node1 = dummy1, node2 = dummy2, curr = head;
        while (curr != null) {
            if (curr.val < x) {
                node1.next = curr;
                node1 = node1.next;
            } else {
                node2.next = curr;
                node2 = node2.next;
            }
            curr = curr.next;
        }
        node1.next = dummy2.next;
        node2.next = null;
        return dummy1.next;
    }
}